File: backend/msreserva/src/main/java/com/dac/msreserva/cqrs/ConsultaListener.java
package com.dac.msreserva.cqrs;

import org.modelmapper.ModelMapper;
import org.modelmapper.TypeToken;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.dac.msreserva.DTO.UpdateEstadoDTO;
import com.dac.msreserva.model.ReservaConsulta;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

@Service
public class ConsultaListener {

    private final AsyncReservaService service;
    @Autowired
    private ObjectMapper objectMapper;

    public ConsultaListener(AsyncReservaService service) {
        this.service = service;

    }

    @RabbitListener(queues = "cqrs.gravacao")
    public void gravarReserva(String payload) {
        try {
            System.out.println("CQRS escutado");

            ReservaConsulta reserva = objectMapper.readValue(payload, ReservaConsulta.class);
            service.gravarReserva(reserva);

        } catch (Exception e) {
            System.out.println("Erro ao processar mensagem CQRS: " + e.getMessage());
            e.printStackTrace();
        }
    }

    @RabbitListener(queues = "cqrs.edicao")
    public void editaReserva(String payload) {
        try {
            System.out.println("CQRS pra edição escutado");
            System.out.println(payload);
            UpdateEstadoDTO reserva = objectMapper.readValue(payload, UpdateEstadoDTO.class);
            service.editarReserva(reserva);
        } catch (JsonProcessingException e) {
            System.err.println("Erro ao desserializar payload: " + e.getMessage());
        }
    }

}

File: backend/reserva-service/src/main/kotlin/br/ufpr/dac/reserva_service/resource/cqrs/ConsultaListener.kt
package br.ufpr.dac.reserva_service.resource.cqrs

import br.ufpr.dac.reserva_service.resource.dto.ReservaConsultaInputDTO
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import org.springframework.amqp.rabbit.annotation.RabbitListener
import org.springframework.stereotype.Service
import utils.dto.ReservaUpdateEstadoDTO
import utils.gson.ZonedDateTimeAdapter
import java.time.ZonedDateTime

@Service
class ConsultaListener(private val service: AsyncReservaService) {
    private final val gson: Gson = GsonBuilder()
        .registerTypeAdapter(ZonedDateTime::class.java, ZonedDateTimeAdapter())
        .create()

    @RabbitListener(queues = ["emiratads.cqrs.gravacao"])
    fun gravarReserva(payload: String) {
        val reservas: List<ReservaConsultaInputDTO> = gson.fromJson(
            payload,
            object : TypeToken<List<ReservaConsultaInputDTO>>() {}.type
        )
        service.gravarReserva(reservas)
    }

    @RabbitListener(queues = ["emiratads.cqrs.edicao"])
    fun editaReserva(payload: String) {
        val reservas = gson.fromJson(payload, ReservaUpdateEstadoDTO::class.java)
        service.editarReserva(reservas)
    }
}